decision_service/include/object_pool.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include <mutex>
+#include <vector>
+
+namespace reinforcement_learning {
+  template<typename TObject>
+  class pooled_object {
+  private:
+    TObject * _val;
+
+  public:
+    pooled_object(TObject* obj, int pversion)
+      : _val(obj), version(pversion)
+    { }
+
+    ~pooled_object()
+    {
+      if (_val) {
+        delete _val;
+        _val = nullptr;
+      }
+    }
+
+    inline TObject* val() { return _val; }
+
+    const int version;
+  };
+
+  template<typename TObject, typename TFactory>
+  class object_pool;
+
+  // RAII guard to handle exception case properly
+  template<typename TObject, typename TFactory>
+  class pooled_object_guard {
+    object_pool<TObject, TFactory>* _pool;
+    pooled_object<TObject>* _obj;
+
+  public:
+    pooled_object_guard(object_pool<TObject, TFactory>& pool, pooled_object<TObject>* obj)
+      : _pool(&pool), _obj(obj)
+    { }
+
+    ~pooled_object_guard() {
+      _pool->return_to_pool(_obj);
+    }
+
+    TObject* operator->() { return _obj->val(); }
+  };
+
+  template<typename TObject, typename TFactory>
+  class object_pool {
+    int _version;
+    std::vector<pooled_object<TObject>*> _pool;
+    TFactory* _factory;
+    std::mutex _mutex;
+    int _used_objects;
+
+  public:
+    object_pool(TFactory* factory)
+      : _version(0), _factory(factory), _used_objects(0)
+    { }
+
+    ~object_pool() {
+      std::lock_guard<std::mutex> lock(_mutex);
+
+      // delete factory
+      if (_factory) {
```

Comment:
>if (_factory) { [](start = 6, length = 15)

need to either a) delete copy const, assignment + moves or b) implement them to prevent accidental misuse